### PR TITLE
Issue #2 - use only async node creation in AbstractCloudProviderTest to speed up the tests

### LINKED ISSUE #2: Use only the async node creation in AbstractCloudProviderTest
Creating the nodes only once by `createNodes` will save significant amount of time for running the providers smoke test.

### PR DESCRIPTION
**Pull request by kwart**
Fixes #2

### CODE REVIEW

File: core/src/test/java/org/wildfly/extras/sunstone/tests/AbstractCloudProviderTest.java [hunk 1]
```diff
@@ -346,23 +327,4 @@ private void testFileCopying(Node node) throws IOException, InterruptedException
...
-        if (isNullOrEmpty(string)) {
-            fail("Expected not-null nor empty string: '" + string + "'");
-        }
-    }
-
-    private static void assertNullOrEmpty(String string) {
-        if (!isNullOrEmpty(string)) {
-            fail("Expected null or empty string: '" + string + "'");
-        }
-    }
```
Comment (Ladicek) [inline core/src/test/java/org/wildfly/extras/sunstone/tests/AbstractCloudProviderTest.java]: Thanks for this :-)

File: core/src/test/java/org/wildfly/extras/sunstone/tests/AbstractCloudProviderTest.java [hunk 2]
```diff
@@ -155,19 +157,28 @@ public final void node() throws Exception {
                 // OK
             }
 
-            try (Node node = cloudProvider.createNode("mynode")) {
-                assertEquals(1, cloudProvider.getNodes().size());
-                assertNotNull(cloudProvider.getNode("mynode"));
+            try (CreatedNodes createdNodes = cloudProvider.createNodes("mynode", "othernode")) {
```
Comment (Ladicek) [inline core/src/test/java/org/wildfly/extras/sunstone/tests/AbstractCloudProviderTest.java]: This was originally to test the behavior of the cloud provider when nodes are incrementally being started and stopped. I'd like to keep that part of the code covered, but I understand that performance is a big concern.

How about we only do this test with one node (the "othernode" is really not relevant here and can be removed), and use some kind of mocking for testing how the cloud provider behaves when nodes are started and stopped?

Comment (kwart) [inline core/src/test/java/org/wildfly/extras/sunstone/tests/AbstractCloudProviderTest.java]: IMO, this parallel 2 nodes start gives nearly the same coverage as when started one by one, but with great performance gain. The logic behind `createNodes` is calling `createNode` so we only lost the very small part of assertions.

If we use only one node test, then the coverage is significantly lower.

**Comment by LittleJohnII** [comment]
LGTM.

I tend to agree with @kwart on the incremental node starting. But maybe we could do `CompleteableFuture<Node> cf = cloudProvider.createNodeAsync("thirdNode")` after the first two nodes are started and `cf.join()` after all the file copying, lifecycle control, etc. tests are done? That way the performance drop wouldn't be that big and we would still have incremental starting covered.

File: core/src/test/java/org/wildfly/extras/sunstone/tests/AbstractCloudProviderTest.java (hunk 2 above)
Comment (Ladicek) [inline core/src/test/java/org/wildfly/extras/sunstone/tests/AbstractCloudProviderTest.java]: I don't think that removing "othernode" would lower test coverage. A collection of one is not that different from a collection of two, if they are used the same way.

Anyway, if you're fine with the test's performance right now, with two nodes, I'm OK with that too :-)

**Comment by LittleJohnII** [comment]
Unless someone has something to add, I'll assume you've reached agreement and I'll merge this.

**Comment by Ladicek** [comment]
With a bit of hesitation, +1 :-)

**Pull request merged**